src/modules/rlm_magic/rlm_magic.c

```diff
@@ -0,0 +1,249 @@
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/**
+ * $Id$
+ * @file rlm_magic.c
+ * @brief Example module code.
+ *
+ * @copyright 2013 The FreeRADIUS server project
+ * @copyright 2013 your name (your name\@address)
+ */
+RCSID("$Id$")
+
+#include <freeradius-devel/radiusd.h>
+#include <freeradius-devel/modules.h>
+#include <freeradius-devel/rad_assert.h>
+
+/*
+ *	Define a structure for our module configuration.
+ *
+ *	These variables do not need to be in a structure, but it's
+ *	a lot cleaner to do so, and a pointer to the structure can
+ *	be used as the instance handle.
+ */
+typedef struct
+{
+	unsigned short port;
+	char const *socket_address;
+	bool is_local;
+} rlm_magic_t;
+
+/*
+ *	A mapping of configuration file names to internal variables.
+ */
+static const CONF_PARSER module_config[] = {
+	{"port", FR_CONF_OFFSET(PW_TYPE_SHORT, rlm_magic_t, port), "12345"},
+	{"is_local", FR_CONF_OFFSET(PW_TYPE_BOOLEAN, rlm_magic_t, is_local), false},
+	{"socket-address", FR_CONF_OFFSET(PW_TYPE_STRING, rlm_magic_t, socket_address), "127.0.0.1"},
+	CONF_PARSER_TERMINATOR};
+
+/*
+ *	Do any per-module initialization that is separate to each
+ *	configured instance of the module.  e.g. set up connections
+ *	to external databases, read configuration files, set up
+ *	dictionary entries, etc.
+ */
+static int mod_instantiate(UNUSED CONF_SECTION *conf, void *instance)
+{
+	rlm_magic_t *inst = instance;
+	DEBUG2("Creating Magic module");
+
+	if (inst->is_local)
+	{
+		#ifndef HAVE_SYS_UN_H
+			ERROR("Trying to use local unix sockets on a system that does not support it");
+			return -1;
+		#endif
+		DEBUG2("Using local Unix socket %s for authentication", inst->socket_address);
+	}
+	else
+	{
+		DEBUG2("Using remote server %s:%i for authentication", inst->socket_address, inst->port);
+	}
+
+	return 0;
+}
+
+/*
+ *	Don't currently do anything here
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authorize(UNUSED void *instance, UNUSED REQUEST *request)
+{
+	return RLM_MODULE_NOOP;
+}
+
+/*
+ *	Authenticate the user with the given password.
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(void *instance, REQUEST *request)
+{
+	// The instance variables
+	rlm_magic_t *inst = instance;
+
+	//Socket related vars
+	char buffer[BUFSIZ];
+	int sockfd;
+
+	DEBUG2("Attempting Magic Authentication");
+
+	if (inst->is_local)
+	{
+		sockfd = fr_socket_client_unix(inst->socket_address, false);
+		if (sockfd == -1)
+		{
+			fr_perror("socket");
+			return RLM_MODULE_FAIL;
+		}
+	}
+	else
+	{
+		fr_ipaddr_t address_info;
+		address_info.af = AF_INET;
+		struct hostent *hostent;
+		
+		hostent = gethostbyname(inst->socket_address);
```

Comment:
This is a non-starter for run-time use.  DNS lookups should be done when the server starts up, and cached.  If people change DNS while the server is running, too bad.

The alternative is to make the server dependent on DNS for *every single packet*.  So if DNS goes down, this line will *lock* the server for 30 seconds.  And very bad things will happen.

Some modules do allow hostnames for server identifiers.  This is generally done only when the third-party APIs don't really support IP addresses.

Reply:
makes sense, this would then happen in the mod_instantiate function correct?

Reply:
Yes, the instantiate module should do that lookup and cache the IP address in the instance data.  There's a `fr_pton()` function which handles all of the DNS lookups, and IPv4 / IPv6 issues.  The `gethostbyname()` API is just not useful.